File: mkdocs/snippets/devbook/network-currency/query_block_rewards.py
import json
import os
import urllib.request

from symbolchain.CryptoTypes import PublicKey
from symbolchain.facade.NemFacade import NemFacade

NODE_URL = os.getenv('NODE_URL', 'http://libertalia.nemtest.net:7890')
print(f'Using node {NODE_URL}')

BLOCK_HEIGHT = os.getenv('BLOCK_HEIGHT', '661258')

facade = NemFacade('testnet')

try:
	# Fetch the block at the given height [>step-1]
	block_url = f'{NODE_URL}/block/at/public'
	request = urllib.request.Request(
		block_url,
		data=json.dumps({'height': int(BLOCK_HEIGHT)}).encode(),
		headers={'Content-Type': 'application/json'})
	with urllib.request.urlopen(request) as response:
		block = json.loads(response.read())
	transactions = block['transactions']
	print(f'Block height: {BLOCK_HEIGHT}')
	print(f'Transactions: {len(transactions)}')
	# [<step-1]
	# Identify the harvester [>step-2]
	harvester = facade.network.public_key_to_address(
		PublicKey(block['signer']))
	print(f'Harvester: {harvester}')
	# [<step-2]
	# Sum the transaction fees [>step-3]
	total_reward = 0
	print('\nTransaction fees:')
	for transaction in transactions:
		fee = int(transaction['fee'])
		total_reward += fee
		print(f'  Fee: {fee / 1e6:,.6f} XEM')
	# [<step-3]
	# Total reward [>step-4]
	print(f'\nTotal block reward: {total_reward / 1e6:,.6f} XEM')
	# [<step-4]
except Exception as error:
	print(error)

File: mkdocs/snippets/devbook/network-currency/QueryBlockRewards.java
//JAVA 21+
//DEPS org.symbol:symbol-sdk:3.3.3

import java.io.IOException;
import java.math.BigInteger;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.net.http.HttpResponse.BodyHandlers;
import java.util.Locale;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;

import org.symbol.sdk.CryptoTypes;
import org.symbol.sdk.facade.NemFacade;
import org.symbol.sdk.nem.Address;

public final class QueryBlockRewards {
	private static final HttpClient HTTP_CLIENT =
		HttpClient.newHttpClient();

	private static final ObjectMapper JSON_MAPPER = new ObjectMapper();

	private static final String NODE_URL = System.getenv().getOrDefault(
		"NODE_URL", "http://libertalia.nemtest.net:7890");

	private static final String BLOCK_HEIGHT =
		System.getenv().getOrDefault("BLOCK_HEIGHT", "661258");

	// Format an atomic amount as whole XEM with integer math,
	// since amounts can exceed float precision.
	private static String fmt(final BigInteger value) {
		final BigInteger divisor = BigInteger.valueOf(1_000_000);
		final BigInteger[] parts = value.divideAndRemainder(divisor);
		return String.format(Locale.US, "%,d.%06d", parts[0], parts[1]);
	}

	public static void main(final String[] args) {
		try {
			new QueryBlockRewards().run();
		} catch (final Exception ex) {
			System.out.println(null == ex.getMessage()
				? ex.toString()
				: ex.getMessage());
		}
	}

	private void run() throws IOException, InterruptedException {
		System.out.printf("Using node %s%n", NODE_URL);
		final NemFacade facade = new NemFacade("testnet");

		// Fetch the block at the given height [>step-1]
		final ObjectNode requestBody = JSON_MAPPER.createObjectNode();
		requestBody.put("height", Long.parseLong(BLOCK_HEIGHT));
		final HttpRequest request = HttpRequest.newBuilder(
			URI.create(NODE_URL + "/block/at/public"))
			.header("Content-Type", "application/json")
			.POST(HttpRequest.BodyPublishers.ofString(
				requestBody.toString()))
			.build();
		final HttpResponse<String> response = HTTP_CLIENT.send(
			request, BodyHandlers.ofString());
		if (200 != response.statusCode())
			throw new IOException("HTTP error! status: "
				+ response.statusCode());
		final JsonNode block = JSON_MAPPER.readTree(response.body());
		final JsonNode transactions = block.get("transactions");
		System.out.printf("Block height: %s%n", BLOCK_HEIGHT);
		System.out.printf("Transactions: %d%n", transactions.size());
		// [<step-1]
		// Identify the harvester [>step-2]
		final Address harvester = facade.network.publicKeyToAddress(
			new CryptoTypes.PublicKey(block.get("signer").asText()));
		System.out.printf("Harvester: %s%n", harvester);
		// [<step-2]
		// Sum the transaction fees [>step-3]
		BigInteger totalReward = BigInteger.ZERO;
		System.out.println("\nTransaction fees:");
		for (final JsonNode transaction : transactions) {
			final BigInteger fee = new BigInteger(
				transaction.get("fee").asText());
			totalReward = totalReward.add(fee);
			System.out.printf("  Fee: %s XEM%n", fmt(fee));
		}
		// [<step-3]
		// Total reward [>step-4]
		System.out.printf("\nTotal block reward: %s XEM%n",
			fmt(totalReward));
		// [<step-4]
	}
}
